app/src/utils/routing.ts

```diff
@@ -0,0 +1,221 @@
+import { UrlObject } from 'url'
+
+import { SessionData } from '@src/types'
+import {
+  isValidChooseClinic,
+  isValidContact,
+  isValidEligibility,
+  isValidIncome,
+} from '@utils/dataValidation'
+
+interface RestrictedPages {
+  [key: string]: string[]
+}
+
+const pageFlow = [
+  '/',
+  '/how-it-works',
+  '/eligibility',
+  '/income',
+  '/choose-clinic',
+  '/contact',
+  '/review',
+  '/confirmation',
+]
+
+export function getForwardRoute(
+  pathname: string,
+  reviewMode: boolean,
+  session: SessionData | ((value: SessionData) => void)
+): UrlObject | string {
+  const position = pageFlow.indexOf(pathname)
+
+  // Check for simple review mode cases.
+  const reviewModePages = ['/income', '/choose-clinic', '/contact']
+  if (reviewMode && reviewModePages.includes(pathname)) {
+    return '/review'
+  }
+
+  // Check for edge cases.
+  // /other-benefits always routes forward to /
+  if (pathname === '/other-benefits') {
+    return '/'
+  }
+  // There is no action button on /confirmation, so return empty string.
+  else if (pathname === '/confirmation') {
+    return ''
+  }
+  // /eligibility has different behaviour depending on user data.
+  else if (pathname === '/eligibility') {
+    // Typescript believes it's possible for session to be the function:
+    // `(value: SessionData) => void`. If this actually happens at runtime, we
+    // throw an error.
+    if (typeof session === 'function') {
+      throw new Error(
+        'Forward route error: expected a session, but none was found'
+      )
+    }
+    // Otherwise, we can do actual checks against user data to get the correct route.
+    else {
+      if (
+        !isValidEligibility(session.eligibility) ||
+        session.eligibility.residential === 'no' ||
+        session.eligibility.categorical.includes('none')
+      ) {
+        return '/other-benefits'
+      } else if (session.eligibility.adjunctive.includes('none')) {
+        if (reviewMode) {
+          if (!isValidIncome(session.income)) {
+            return { pathname: '/income', query: { mode: 'review' } }
+          } else {
+            return '/review'
+          }
+        } else {
+          return '/income'
+        }
+      } else {
+        if (reviewMode) {
+          return '/review'
+        } else {
+          return '/choose-clinic'
+        }
+      }
+    }
+  }
+
+  // Otherwise handle simple cases.
+  if (position !== -1) {
+    return pageFlow[position + 1]
+  }
+  // Unknown page! It probably doesn't have an action button.
+  else {
+    return ''
```

Comment:
Note to self: This one might be better if the user gets routed to /eligibility instead of page refresh.